lib/tensorflow/graph.rb

```diff
@@ -86,18 +87,45 @@ def define_op(opName, name , input, device, attrs)
     node
   end
 
+  #
+  # Creates a constant Tensor that is added to the graph with a specified name.
+  # Official documentation of {tf.constant}[https://www.tensorflow.org/versions/r0.9/api_docs/python/constant_op.html#constant].
+  #
+  def constant(name, data, type)
+    tensor = Tensorflow::Tensor.new(data, type)
+    self.constants = {name => tensor}
+    self.define_op("Const", name, nil, "", {"dtype" => tensor.type_num, "value" => tensor})
+  end
+
   def make_attr_value(attribute_type, value)
-      # TODO -> Add support for all types
-      result = nil
-      if attribute_type == "T"
-        result = Tensorflow::AttrValue.new(type: value)
+    result = nil
+    case attribute_type
+    when "type"
+      result = Tensorflow::AttrValue.new(type: value) # just put the value number here
+    when "tensor"
+      tensor_element_type = value.type_num
+      case value.type_num
+      when Tensorflow::TF_DOUBLE
+        result = Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("d*")))
+      when Tensorflow::TF_INT32
+        result = Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("l*")))
+      when Tensorflow::TF_INT64
+        result = Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("q*")))
+      when Tensorflow::TF_COMPLEX128
+        tensor_narray = NArray.complex(value.flatten.length)
+        (0..value.flatten.length - 1).each do |i|
+          tensor_narray[i] = value.flatten[i]
+        end
+        result = Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: tensor_narray.to_s))
       end
-      result
+    end
+    result
```

Comment:
First, you don't need `result` variable at all: all branches just finish with calculation of result, so it will be returned in this case:
```ruby
   case attribute_type
    when "type"
      Tensorflow::AttrValue.new(type: value) # just put the value number here
    when "tensor"
      tensor_element_type = value.type_num
      case value.type_num
      when Tensorflow::TF_DOUBLE
        Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("d*")))
      when Tensorflow::TF_INT32
        Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("l*")))
      when Tensorflow::TF_INT64
        Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: value.flatten.pack("q*")))
      when Tensorflow::TF_COMPLEX128
        tensor_narray = NArray.complex(value.flatten.length)
        (0..value.flatten.length - 1).each do |i|
          tensor_narray[i] = value.flatten[i]
        end
        Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: tensor_narray.to_s))
      end
    end
```

But it seems code could be simplified even more: there is exactly the same lines in all `"tensor"` branch:
```ruby
Tensorflow::AttrValue.new(tensor: Tensorflow::TensorProto.new(dtype: value.type_num, tensor_shape: value.tensor_shape_proto, tensor_content: "the only difference is here"))
```
So, what do you think about this:
```ruby
    case attribute_type
    when "type"
      Tensorflow::AttrValue.new(type: value) # just put the value number here
    when "tensor"
      tensor_element_type = value.type_num
      content =
        case value.type_num
        when Tensorflow::TF_DOUBLE
          value.flatten.pack("d*")
        when Tensorflow::TF_INT32
          value.flatten.pack("l*")
        when Tensorflow::TF_INT64
          value.flatten.pack("q*")
        when Tensorflow::TF_COMPLEX128
          tensor_narray = NArray.complex(value.flatten.length)
          (0..value.flatten.length - 1).each do |i|
            tensor_narray[i] = value.flatten[i]
          end
          tensor_narray.to_s
        end
      Tensorflow::AttrValue.new(
        tensor: Tensorflow::TensorProto.new(
          dtype: value.type_num,
          tensor_shape: value.tensor_shape_proto,
          tensor_content: content
        )
      )
    end
```
Now we can see that `content =` calculation could be easily extracted into separate method, making code to look somewhat like this:
```ruby
   case attribute_type
    when "type"
      Tensorflow::AttrValue.new(type: value) # just put the value number here
    when "tensor"
      Tensorflow::AttrValue.new(
        tensor: Tensorflow::TensorProto.new(
          dtype: value.type_num,
          tensor_shape: value.tensor_shape_proto,
          tensor_content: value_to_tensor_content(value)
        )
      )
    end
```

And, finally, typically it is a good practice to have `else` branch in `case` (even if you are sure it is impossible to reach -- impossibility handler, raising an error, could save you a hours of debugging).